### ISSUE TITLE
[C#] Implement new Concept Exercise: overflows

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/v3#1138]
This issue describes how to implement the `overflows` concept exercise for the C# track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Goal

The goal of this exercise is to teach the student the basics of the Concept of Overflows in [C#].

## Learning objectives

- Explain integral number overflows.
- Explain floating-point number overflows.
- Know how to use `checked` and `unchecked` to deal with overflows.

## Out of scope

- BigInteger

## Concepts

This Concepts Exercise's Concepts are:

- `overflows`: explain integral number overflows; explain floating-point number overflows; know how to use `checked` and `unchecked` to deal with overflows.

## Prequisites

This Concept Exercise's prerequisites Concepts are:

- `integral-numbers`: know about the integral number types.
- `floating-point-numbers`: know about the floating-point number types.
- `exceptions`: explain how an `OverflowException` is thrown when using checked arithmetic.

### Hints

- [Integral numeric types][integral-numeric-types]: overview of the integral numeric types.
- [Floating-point numeric types][floating-point-numeric-types]: overview of the floating-point numeric types.
- [Numeric conversions][numeric-conversions]: overview of implicit and explicit numeric conversions.
- [Checked and unchecked arithmetic][checked-and-unchecked]: introduction to overflows

### After

- [Integral numeric types][integral-numeric-types]: overview of the integral numeric types.
- [Floating-point numeric types][floating-point-numeric-types]: overview of the floating-point numeric types.
- [Numeric conversions][numeric-conversions]: overview of implicit and explicit numeric conversions.
- [Checked and unchecked arithmetic][checked-and-unchecked]: introduction to overflows
- [`checked` keyword][checked-keyword]: `checked` keyword reference.
- [`unchecked` keyword][unchecked-keyword]: `unchecked` keyword reference.
- [Computerphile: How Gangnam Style broke YouTube][computerphile-gangnam-style]

## Representer

This exercise does not require any specific representation logic to be added to the [representer][representer].

## Analyzer

This exercise does not require any specific analyzer logic to be added to the [analyzer][analyzer].

## Implementing

To implement this exercise, please [follow these instructions](https://github.com/exercism/v3/blob/master/languages/csharp/reference/implementing-a-concept-exercise.md).

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

[how-to-implement-a-concept-exercise]: https://github.com/exercism/v3/blob/master/docs/maintainers/generic-how-to-implement-a-concept-exercise.md
[implemented-exercises]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/README.md#implemented-exercises
[reference]: https://github.com/exercism/v3/blob/master/languages/csharp/reference/README.md#reference-docs
[reference-char]: https://github.com/exercism/v3/blob/master/reference/types/char.md
[reference-example]: https://github.com/exercism/v3/blob/master/reference/types/string.md#implementations
[analyzer]: https://github.com/exercism/csharp-analyzer
[representer]: https://github.com/exercism/csharp-representer
[exercise-example]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/numbers-floating-point
[design-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/design.md
[config.json-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/config.json
[concept-exercises]: https://github.com/exercism/v3/blob/master/docs/concept-exercises.md
[computerphile-gangnam-style]: https://www.youtube.com/watch?v=vA0Rl6Ne5C8
[integral-numeric-types]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/integral-numeric-types
[floating-point-numeric-types]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/floating-point-numeric-types
[numeric-conversions]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/numeric-conversions
[checked-and-unchecked]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/checked-and-unchecked
[checked-keyword]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/checked
[unchecked-keyword]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/unchecked

**Comment by mikedamay** [comment]
Please have a look at proposed draft [instructions](https://github.com/mikedamay/v3/blob/csharp/overflow/languages/csharp/exercises/concept/overflow/.docs/instructions.md) and [code](https://github.com/mikedamay/v3/blob/csharp/overflow/languages/csharp/exercises/concept/overflow/.meta/Example.cs).

I'm proposing to move the `checked` concept to the user-defined exceptions exercise.  It is difficult to have both integers and floating point with the calculator theme.

Let me know what you think.

**Comment by ErikSchierboom** [comment]
> I'm proposing to move the checked concept to the user-defined exceptions exercise. It is difficult to have both integers and floating point with the calculator theme.

I'm not sure about this to be honest. To me, the most interesting aspects of overflows are the checked/unchecked semantics. I do agree that the calculator theme might be hard, so maybe we should try to come up with another theme?

**Comment by mikedamay** [comment]
> the most interesting aspects of overflows are the checked/unchecked semantics

I agree.  But, in general, it's very difficult to illustrate both floating point and integers naturally in the same story.  If you're using real numbers then everything significant tends to be real numbers, ditto decimals.  The exception is for truncation.  My TI 58 programmable calculator had a function actually called "int" to do truncation so we could implement that.  Other than that, I'm lost for lost for inspiration.

**Comment by ErikSchierboom** [comment]
Maybe we could ask for inspiration on the #v3 channel?

**Comment by mikedamay** [comment]
> Maybe we could ask for inspiration on the #v3 channel?

Will do.  I didn't want to too readily abuse the v3 channel's goodwill after our great success with `structs`.

**Comment by ErikSchierboom** [comment]
Hah. I could also post it there is you'd like?

**Comment by mikedamay** [comment]
> I could also post it there is you'd like?

That's a  good idea.  Thanks.  I suspect my negativity would have come across.

**Comment by ErikSchierboom** [comment]
Done!

**Comment by ErikSchierboom** [comment]
@mikedamay Some ideas have been posted in [this thread on Slack](https://exercism-team.slack.com/archives/CR91YFNG3/p1593587160122600). Personally, I think the hyperinflation story might be nice.

**Comment by mikedamay** [comment]
> the hyperinflation story might be nice.

Agreed we can have notes in integer amounts and GDP in floats.  In reality of course as hyperinflation hits you get very poor, very quickly and probably stop measuring GDP anyway but I think we can get away with it :-).

**Comment by mikedamay** [comment]
closed by PR #1867
